### PR TITLE
:sparkles:(widgets) add widget customization

### PR DESCRIPTION
**Pull request by Kitero**
## Purpose

The lib needs a way to customize the widgets such as required or not.

## Proposal

Giving a new prop of type JSONSchema7, so the user can indicate for every widget what parameters input to render.
The application of the parameter will be proceeded directly inside the library.

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -1,3 +1,7 @@
+import type { UiSchema } from '@rjsf/core';
+
+const SEPARATOR_ID_RJSF = '_';
```
Comment (jbpenrath) [inline src/utils.ts:3]: For maintainability purpose, I think this constant should be passed to the `idSeparator` prop of our global `Form`. In this way, we have the guarantee that this constant is in line with your the RJSF Form settings. We could declare it in a `settings.ts` file or in a `.env` file.

Furthermore I would renamed this constant `RJSF_ID_SEPARATOR`. I find it more readable, in this way all constant relying on react json schema form will be prefixed by `RJSF`

File: src/utils.ts [hunk 2]
```diff
@@ -10,11 +14,19 @@ function makeid(length: number) {
...
 
 function getCurrentField(id: string): string {
-  return id.split('_').pop() || id;
+  return id.split(SEPARATOR_ID_RJSF).pop() || id;
+}
+
+function getTemplateWidgetName(id: string, uiSchema: UiSchema) {
+  const idParts = id.split(SEPARATOR_ID_RJSF);
+  const section = uiSchema?.[idParts[1]];
+  return idParts.length === 3
```
Comment (jbpenrath) [inline src/utils.ts:27]: Add a comment to explain precisely you check `idParts === 3`

File: src/context/VernaContextProvider.tsx [hunk 3]
```diff
@@ -159,16 +164,20 @@ function VernaContextProvider({
...
 
   function handleSubmit<FormData = unknown>(callback?: (formData: unknown) => void) {
-    return (event: ISubmitEvent<FormData>) => {
+    return (event: ISubmitEvent<FormData>, nativeEvent: FormEvent<HTMLFormElement>) => {
       const newFormData = selector ? { ...formData, [selector]: event.formData } : event.formData;
+      nativeEvent.stopPropagation();
+      nativeEvent.preventDefault();
       setFormData(newFormData);
       callback && callback(newFormData);
+      return false;
```
Comment (jbpenrath) [inline src/context/VernaContextProvider.tsx:173]: Test residue nope?

File: src/Widgets/configuration/WidgetParametersApplyer.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+const parametersApplyer: ParameterApplyer = {
+  items: updateItemsValidator,
+  maxLength: updateMaxLengthValidator,
+  required: updateRequiredValidator,
+  // maximum:,
+  // minimum:,
+};
+
+export default function WidgetParametersApplyer(
```
Comment (jbpenrath) [inline src/Widgets/configuration/WidgetParametersApplyer.tsx:99]: Maybe `Modifier` term is better than `Applyer`

```suggestion
export default function modifyWidgetParameters(
```

File: src/Widgets/configuration/WidgetParametersApplyer.tsx [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  verna.setSchema(newSchema);
+}
+
+export interface WidgetParameters {
+  items: Maybe<string[]>;
+  required: Maybe<boolean>;
+  maxLength: Maybe<number>;
+}
+
+type ApplierFunction<T> = (newValue: T, verna: VernaContextProps, widgetIdParts: string[]) => void;
```
Comment (jbpenrath) [inline src/Widgets/configuration/WidgetParametersApplyer.tsx:85]:
```suggestion
type WidgetParameterModifier<T> = (value: T, verna: VernaContextProps, widgetIdParts: string[]) => void;
```

File: src/Widgets/configuration/WidgetParametersApplyer.tsx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+type ParameterApplyer = {
+  [key in keyof Parameters]: ApplierFunction<Parameters[key]>;
+};
```
Comment (jbpenrath) [inline src/Widgets/configuration/WidgetParametersApplyer.tsx:89]:
```
type WidgetParameterModifiers = {
  [key in keyof Parameters]: WidgetParameterModifier<Parameters[key]>;
};
```

**Comment by jbpenrath** [comment]
Do not forget to write tests 😉

File: src/VernaForm.tsx [hunk 7]
```diff
@@ -1,24 +1,32 @@
...
       schema={schema}
       uiSchema={uiSchema}
       formData={selectedFormData}
       onSubmit={handleSubmit(onSubmit)}
       widgets={widgets}
-      FieldTemplate={RenderEditorFieldTemplate}
-    />
+      disabled={isEditor}
+      FieldTemplate={RenderFieldTemplate}
+      liveValidate={!isEditor}
```
[the remark this answers is not in the document]
Comment (Kitero) [reply src/VernaForm.tsx]: J'ai ajouté la regle dans la conf typescript, hésite pas si un des parametres te chagrine.

File: src/Widgets/DropZone.tsx [hunk 8]
```diff
@@ -1,11 +1,15 @@
 import React, { useState } from 'react';
-import type { WidgetProps } from '@rjsf/core';
 import ShowCaseWidgetProps from '../types/Widgets';
 import { makeid } from '../utils';
 import { addItemToSchema, addItemToUiSchema } from '../context/InteractionMethods';
 import { useVerna } from '../context/VernaContextProvider';
 
-export default function DropZone({ id }: Partial<WidgetProps>) {
```
[the remark this answers is not in the document]
Comment (Kitero) [reply src/Widgets/DropZone.tsx:8]: Indeed, it was a working on feature, to open the edit parameter box, so the user can immediatly edit the new field (and set a proper name)

File: src/utils.ts (hunk 1 above)
Comment (Kitero) [reply src/utils.ts:3]: Ohhh good tips !! I'll go with it thanks !

File: src/VernaForm.tsx [hunk 9]
```diff
@@ -1,24 +1,32 @@
...
+    >
+      <button type="submit" name="verna-form">
+        Save
+      </button>
```
[the remark this answers is not in the document]
Comment (Kitero) [reply src/VernaForm.tsx]: Those 3 lines are not that usefull because if i remove it,it's replaces with a default submit button, i only use it for customization purpose, we would need to use a style to hide it, or wrap it in a div or whatever and not display the button if in editor mode.
I searched a bit for it, and there is a uiSchema instruction to customize it, hide and change text, but it seems a bit working on, it works on the playground but can't find how to make it work in verna, but i find a recent commit (13 april), that seems to add it
https://github.com/rjsf-team/react-jsonschema-form/commit/57f00cb522c698a2a201c7204d78dfc6323ac6f6
So i would suggest to wait a bit for the next release to add this fix, we can just disable it for now and do it a clean way once it's release?

nb: i checked the sources of the lib in node_modules, the files managing this truly doesn't exists, with the next release it will be really nice to customize and every client will be able to customize it easily

File: src/context/VernaContextProvider.tsx (hunk 3 above)
Comment (Kitero) [reply src/context/VernaContextProvider.tsx:173]: Bien vu

File: src/utils.ts [hunk 10]
```diff
@@ -14,19 +13,22 @@ function makeid(length: number) {
...
 function getTemplateWidgetName(id: string, uiSchema: UiSchema) {
-  const idParts = id.split(SEPARATOR_ID_RJSF);
+  const idParts = id.split(RJSF_ID_SEPARATOR);
   const section = uiSchema?.[idParts[1]];
+
+  // If the id is made of 3 parts, it's that there is a section before the widget used
+  // Then you need to get it in the UiSchema before accessing the ui:widget
   return idParts.length === 3
     ? section?.[idParts[2]]?.['ui:widget'] || ''
     : section?.['ui:widget'] || '';
```
[the remark this answers is not in the document]
Comment (Kitero) [reply src/utils.ts]: I modified a bit the text & variable names because of the case with a selector, the names are not valid anymore.
Otherwise it's a great change thanks :)

File: examples/playground/src/widgets/SelectWidget.tsx [hunk 11]
```diff
@@ -11,11 +11,13 @@ export default function SelectWidget(props: Partial<WidgetProps>) {
...
+        id={props.id}
         onChange={(event) => props.onChange && props.onChange(event.target.value)}
+        required={props.required}
+        value={props.value}
       >
-        {(props.options?.enumOptions as option[] | undefined)?.map((option, index) => (
+        {((props.options?.enumOptions || []) as option[])?.map((option, index) => (
           <option key={`${option.label}-${index}`} value={option.value}>
             {option.label}
           </option>
```
[the remark this answers is not in the document]
Comment (Kitero) [reply examples/playground/src/widgets/SelectWidget.tsx]: I fixed it :)

File: examples/playground/src/App.tsx [hunk 12]
```diff
@@ -16,6 +16,15 @@ function App() {
...
         properties: {
+          checkboxes: {
+            items: {
+              enum: ['AAAA', 'BBBB', 'CCCC'],
+              type: 'string',
+            },
+            title: 'LALALALA',
+            type: 'array',
+            uniqueItems: true,
+          },
```
[the remark this answers is not in the document]
Comment (Kitero) [reply examples/playground/src/App.tsx]: Ahh yes indeed ahah

File: src/templates.ts [hunk 13]
```diff
@@ -4,10 +4,17 @@ import Section from './widgets/Section';
...
-  title,
-  type,
-});
+): JSONSchema7Definition => {
+  const def: JSONSchema7Definition = {
+    title,
+    type,
+  };
+  if (type === 'array') {
+    def['items'] = { enum: ['AAA'], type: 'string' };
```
[the remark this answers is not in the document]
Comment (Kitero) [reply src/templates.ts:13]: I forgot to remove it, was a testing data :)

File: examples/playground/src/widgets/SelectWidget.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply examples/playground/src/widgets/SelectWidget.tsx]: And what do you think about props destructuring and the use of `defaultProps`to prevent `props.onChange && ...` ?

File: src/utils.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/utils.ts]: I do not see the change

File: src/Widgets/configuration/WidgetParametersConfiguration.tsx [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    const currentWidgetName = widgetIdParts[verna.selector ? 1 : 2];
+    const currentSection = ((verna.selector
+      ? verna.schema
+      : verna.schema.properties?.[widgetIdParts[1]]) || {}) as {
+      [key: string]: JSONSchema7Definition;
+    };
+    const self =
+      ((verna.selector
+        ? currentSection?.[widgetIdParts[1]]
+        : (currentSection as JSONSchema7)?.properties?.[widgetIdParts[2]]) as JSONSchema7) || {};
```
Comment (jbpenrath) [inline src/Widgets/configuration/WidgetParametersConfiguration.tsx]: This seems weird... Line 45 you cast `currentSection` as `{[key:string]: JSONSchema7Definition}` then line 49, you cast the same variable as `JSONSchema7`. I think this hide something is wrong here.

File: src/utils/schema/getters.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+): VernaJSONSchemaType {
+  if (defaultSelector && defaultSchema?.properties?.[defaultSelector]) {
+    return defaultSchema.properties[defaultSelector] as VernaJSONSchemaType;
+  }
+  return defaultSchema;
+}
+
+/**
+ * TODO: explain the purpose of this method
+ *
```
Comment (jbpenrath) [inline src/utils/schema/getters.ts:20]: There is something to write here ^^

File: src/utils/schema/getters.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+  defaultSelector: string | undefined,
+): UiSchema {
+  if (defaultSelector && defaultUiSchema?.[defaultSelector]) {
+    return defaultUiSchema[defaultSelector];
+  }
+  return defaultUiSchema;
+}
+
+/**
+ * TODO: explain the purpose of this method
```
Comment (jbpenrath) [inline src/utils/schema/getters.ts:35]: Ditto. There is something to write here ^^

File: src/utils/schema/updateWidgetProperties.ts [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+  enum: Maybe<string[]>;
+  required: Maybe<boolean>;
+  maxLength: Maybe<number>;
+  maximum: Maybe<number>;
+  minimum: Maybe<number>;
+};
+type PropertyName = Properties[keyof Properties];
+
+/**
+ * The WidgetParameterModifier type is the prototype used to apply any
```
Comment (jbpenrath) [inline src/utils/schema/updateWidgetProperties.ts:24]:
```suggestion
 * The PropertyUpdater type is the prototype used to apply any
```

File: src/utils/schema/widgetPropertyUpdaters.ts [hunk 18]
```diff
@@ -0,0 +1,122 @@
+import { VernaContextProps } from '../../providers/VernaProvider';
+import VernaJSONSchemaType from '../../types/rjsf';
+import type { Properties } from './updateWidgetProperties';
+
+/**
+ * This function modify a property on the level of the widget directly inside the Schema
```
Comment (jbpenrath) [inline src/utils/schema/widgetPropertyUpdaters.ts:6]:
```suggestion
 * Update the schema to update a widget property
```

File: src/utils/schema/widgetPropertyUpdaters.ts [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+ */
+function updateProperty(
+  propertyKey: keyof VernaJSONSchemaType,
+  value: unknown,
+  verna: VernaContextProps,
+  widgetPath: string[],
+) {
+  const newSchema = { ...verna.schema };
+  if (verna.selector) {
+    // TODO
```
Comment (jbpenrath) [inline src/utils/schema/widgetPropertyUpdaters.ts:21]: There is something to implement here.

File: src/utils/schema/widgetPropertyUpdaters.ts [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+ * @param widgetPath The decomposed id for the corresponding widget
+ */
+function updateRequired(
+  value: Properties['required'],
+  verna: VernaContextProps,
+  widgetPath: string[],
+) {
+  const newSchema = { ...verna.schema };
+  if (verna.selector) {
+    // TODO
```
Comment (jbpenrath) [inline src/utils/schema/widgetPropertyUpdaters.ts:47]: Ditto

File: src/utils/schema/widgetPropertyUpdaters.ts [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+ * (e.g: a Select field)
+ *
+ * @param value The value to add or removed from required property
+ * @param verna The verna context, used to update the rendered schema
+ * @param widgetPath The decomposed id for the corresponding widget
+ */
+function updateItems(value: Properties['items'], verna: VernaContextProps, widgetPath: string[]) {
+  const newSchema = verna.schema;
+  if (verna.selector) {
+    // TODO
```
Comment (jbpenrath) [inline src/utils/schema/widgetPropertyUpdaters.ts:73]: Ditto

File: examples/playground/src/App.tsx [hunk 22]
```diff
@@ -1,47 +1,175 @@
...
+      },
+    },
     title: 'A registration form',
     type: 'object',
   };
 
   const formData = {};
 
   const widgets = {
+    SelectWidget: SelectWidget,
```
Comment (jbpenrath) [inline examples/playground/src/App.tsx:36]:
```suggestion
    selectWidget: SelectWidget,
```

File: examples/playground/src/App.tsx [hunk 23]
```diff
@@ -1,47 +1,175 @@
...
 
   // TODO: Add management for ui:ObjectFieldTemplate saving
-  const uiSchema: UiSchema = {};
+  const uiSchema: UiSchema = {
+    section: {
+      onlyNumbersString: {
+        'ui:widget': 'textWidget',
+      },
+      select: {
+        'ui:widget': 'SelectWidget',
```
Comment (jbpenrath) [inline examples/playground/src/App.tsx:50]:
```suggestion
        'ui:widget': 'selectWidget',
```

File: examples/playground/src/styles/verna.scss [hunk 24]
```diff
@@ -38,3 +38,96 @@
...
   margin: 5px 0 0 5px;
 }
+
+.widget-parameters-wrapper {
+  width: 350px;
+  padding: 8px;
+  background-color: #b9b9b9;
+  border: solid 1px gray;
+  display: flex;
+  flex-direction: column;
```
Comment (jbpenrath) [inline examples/playground/src/styles/verna.scss:48]: Sort lines alphabetically

File: src/components/WidgetParametersForm/index.tsx [hunk 25]
```diff
@@ -0,0 +1,71 @@
+import Form, { type ISubmitEvent, type UiSchema } from '@rjsf/core';
```
[the remark this answers is not in the document]
Comment (Kitero) [reply src/components/WidgetParametersForm/index.tsx:1]: Ahh bien vu

File: src/utils/schema/getters.ts (hunk 15 above)
Comment (Kitero) [reply src/utils/schema/getters.ts:20]: My bad I missed this file

File: src/utils/schema/widgetPropertyUpdaters.ts (hunk 18 above)
Comment (Kitero) [reply src/utils/schema/widgetPropertyUpdaters.ts:6]: Way better, thanks

File: examples/playground/src/App.tsx (hunk 22 above)
Comment (Kitero) [reply examples/playground/src/App.tsx:36]: Just for information, the default widgets of the lib are named with a capital letter, so SelectWidget would override the default SelectWidget

File: src/utils/schema/getters.ts [hunk 26]
```diff
@@ -0,0 +1,41 @@
+import type { UiSchema } from '@rjsf/core';
+import VernaJSONSchemaType from '../../types/rjsf';
+
+/**
+ * Set the JSONSchema used by the library, if there is a selector specified it will return only the
```
Comment (jbpenrath) [inline src/utils/schema/getters.ts:5]:
```suggestion
 * Get the JSONSchema used by the library, if there is a selector specified it will return only the
```

File: src/utils/schema/getters.ts [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+  defaultSelector: string | undefined,
+): VernaJSONSchemaType {
+  if (defaultSelector && defaultSchema?.properties?.[defaultSelector]) {
+    return defaultSchema.properties[defaultSelector] as VernaJSONSchemaType;
+  }
+  return defaultSchema;
+}
+
+/**
+ * Set the UiSchema used by the library, if there is a selector specified it will return only the
```
Comment (jbpenrath) [inline src/utils/schema/getters.ts:19]:
```suggestion
 * Get the UiSchema used by the library, if there is a selector specified it will return only the
```

File: src/utils/schema/getters.ts [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+  defaultSelector: string | undefined,
+): UiSchema {
+  if (defaultSelector && defaultUiSchema?.[defaultSelector]) {
+    return defaultUiSchema[defaultSelector];
+  }
+  return defaultUiSchema;
+}
+
+/**
+ * Set the default values used by the library, if there is a selector specified it will
```
Comment (jbpenrath) [inline src/utils/schema/getters.ts:33]:
```suggestion
 * Get the default values used by the library, if there is a selector specified it will
```

File: src/utils/schema/widgetPropertyUpdaters.ts [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+  } else {
+    const section = newSchema.properties?.[widgetPath[0]];
+    if (section) {
+      const widget = section.properties?.[widgetPath[1]];
+      if (widget && widget.items) {
+        (widget.items as VernaJSONSchemaType).enum = value as string[];
+      }
+    }
+  }
+  verna.setSchema(newSchema);
```
Comment (jbpenrath) [inline src/utils/schema/widgetPropertyUpdaters.ts:97]: What do you think to write things like this to don't repeat yourself. It could be applied to function above.

```suggestion
  const newSchema = verna.schema;
  let widget;

  if (verna.selector) {
    widget = newSchema.properties?.[widgetPath[0]];
  } else {
    const section = newSchema.properties?.[widgetPath[0]];
    widget = section?.properties?.[widgetPath[1]];
  }
  
    if (widget && widget.items) {
      (widget.items as VernaJSONSchemaType).enum = value as string[];
      verna.setSchema(newSchema);
    }
```

File: examples/playground/src/App.tsx (hunk 22 above)
Comment (jbpenrath) [reply examples/playground/src/App.tsx]: Oh ok! And does other widgets start with a lower case letter ? Because if not, maybe the right solution is to use an uppercase for all widgets.

```
const widgets = {
    numberWidget: NumberWidget,
    passwordWidget: PasswordWidget,
    quizWidget: QuizWidget,
    selectWidget: SelectWidget,
    textWidget: TextWidget,
  };
```

**Comment by Kitero** [comment]
Existing widgets only starts with capital letters, but it may avoid problems of mistaking existing ones with the ones we want to add

File: examples/playground/package.json [hunk 30]
```diff
@@ -15,6 +15,9 @@
     "@types/react-dom": "17.0.14",
     "@vitejs/plugin-react": "1.3.0",
     "sass": "1.49.9",
+    "stylelint": "^14.8.2",
+    "stylelint-config-standard": "^25.0.0",
+    "stylelint-order": "^5.0.0",
```
Comment (jbpenrath) [inline examples/playground/package.json:20]: Pin dependencies :
```suggestion
    "stylelint": "14.8.2",
    "stylelint-config-standard": "25.0.0",
    "stylelint-order": "5.0.0",
```

File: examples/playground/.stylelintrc.json [hunk 31]
```diff
@@ -0,0 +1,13 @@
...
+    "stylelint-order"
+  ],
+  "rules": {
+    "order/order": [
+      "custom-properties",
+      "declarations"
+    ],
+    "order/properties-alphabetical-order": true
+  }
+}
```
Comment (jbpenrath) [inline examples/playground/.stylelintrc.json:13]: Good idea !
Installation and configuration of stylelint should be in a dedicated commit.

File: src/components/WidgetPropertiesForm/index.tsx [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+      minLength: self.minLength,
+      minimum: self.minimum,
+      required:
+        currentSection.required && (currentSection.required as string[]).includes(widgetName),
+    };
+  }
+
+  return (
+    <Form
+      className="widget-parameters-wrapper"
```
Comment (jbpenrath) [inline src/components/WidgetPropertiesForm/index.tsx:62]: I missed this one 😬 
```suggestion
      className="widget-properties-form"
```

File: src/utils/schema/widgetPropertyUpdaters.ts [hunk 33]
```diff
@@ -0,0 +1,132 @@
...
+  const newSchema = verna.schema;
+  let widget;
+  if (verna.selector) {
+    widget = newSchema.properties?.[widgetPath[0]];
+  } else {
+    const section = newSchema.properties?.[widgetPath[0]];
+    widget = section?.properties?.[widgetPath[1]];
+  }
+
+  if (widget && widget.items) {
```
Comment (jbpenrath) [inline src/utils/schema/widgetPropertyUpdaters.ts:90]:
```suggestion
  if (widget?.items) {
```
